## Homework 5

In [1]:
import numpy as np
import math

#### Exercise 7

In [2]:
def euler(f, a, b, N=10, alpha=0):
    h = (b-a)/N
    t = a
    w = alpha
    points = []
    for i in range(N+1):
        w = w + h*f(t, w)
        t = a + i*h
        points.append((t, w))
    return points

In [3]:
N = 10
f = lambda t, y: 2*y/t + t**2 * math.exp(t)
y = lambda t: t**2 * (math.exp(t) - math.e) 
approximation = np.array(euler(f, 1, 2, N=N, alpha=0))
actual_result = np.array([(1+i*0.1, y(1+i*0.1)) for i in range(N+1)])
errors = abs(approximation - actual_result)
print(f'Approximation: \n{approximation} \n Actual result: \n{actual_result}')
print(f'Errors: \n{errors}')
print(f'Sum of errors: {sum(errors)[1]}')

Approximation: 
[[ 1.          0.27182818]
 [ 1.1         0.598022  ]
 [ 1.2         1.07025736]
 [ 1.3         1.72673043]
 [ 1.4         2.6124924 ]
 [ 1.5         3.78052479]
 [ 1.6         5.29297481]
 [ 1.7         7.22257296]
 [ 1.8         9.65425822]
 [ 1.9        12.68703936]
 [ 2.         16.43612508]] 
 Actual result: 
[[ 1.          0.        ]
 [ 1.1         0.34591988]
 [ 1.2         0.86664254]
 [ 1.3         1.60721508]
 [ 1.4         2.62035955]
 [ 1.5         3.96766629]
 [ 1.6         5.72096153]
 [ 1.7         7.96387348]
 [ 1.8        10.79362466]
 [ 1.9        14.32308154]
 [ 2.         18.68309708]]
Errors: 
[[0.         0.27182818]
 [0.         0.25210213]
 [0.         0.20361483]
 [0.         0.11951535]
 [0.         0.00786715]
 [0.         0.1871415 ]
 [0.         0.42798672]
 [0.         0.74130052]
 [0.         1.13936644]
 [0.         1.63604218]
 [0.         2.246972  ]]
Sum of errors: 7.2337369967194665


In [4]:
"""
    Input: 2 Points a and b in the form (x, y) and the position x at which to interpolate
    Interpolates the point by creating a line between the given points and taking the y-value at x from the line.
"""
def linear_interpolation(a, b, x0):
    m = (b[1] - a[1])/(b[0] - a[0]) # Slope of the line
    b = a[1] - m*a[0] # get intercept from slope and one point
    y = lambda x: m*x + b
    return y(x0)

In [5]:
# i
x0 = 1.04
appr = linear_interpolation(approximation[0], approximation[1], x0)
print(f'Approximation for y({x0})={appr}')

# ii
x0 = 1.55
appr = linear_interpolation(approximation[5], approximation[6], x0)
print(f'Approximation for y({x0})={appr}')

# iii
x0 = 1.97
appr = linear_interpolation(approximation[9], approximation[10], x0)
print(f'Approximation for y({x0})={appr}')

Approximation for y(1.04)=0.40230571061193876
Approximation for y(1.55)=4.53674980081059
Approximation for y(1.97)=15.311399361962671


#### Exercise 8

In [6]:
def taylor_eval(w0, w, N, exact=lambda *args: None):
    res = [w0]
    exact_res = [w0]
    for i in range(1, N):
        res.append(w(res[i-1], i))
        exact_res.append(exact(i))
    return res, exact_res

In [7]:
w = lambda w0, i: w0 + 0.5*( (w0**2 + w0)/(1 + i*0.5) * ( 1 + w0 / (2*(1 + i*0.5)) ) )
exact = lambda i: 2*(1 + i*0.5)/(1 - 2*(1+i*0.5))
result, exact_res = taylor_eval(-2, w, 4, exact=exact)
result = np.array(result)
exact_res = np.array(exact_res)
print(f'Approximated Result:\n{result}\nExact Result:\n{exact_res}\nErrors:\n{abs(result - exact_res)}')

Approximated Result:
[-2.         -1.77777778 -1.58573388 -1.4588846 ]
Exact Result:
[-2.         -1.5        -1.33333333 -1.25      ]
Errors:
[0.         0.27777778 0.25240055 0.2088846 ]


#### Exercise 10

In [8]:
def backward_euler(w0, f, N=10, a=0, b=1, eps=10e-5, iterations=500):
    h = (b-a)/N
    res = [(0, w0)]
    for i in range(1, N+1):
        # Start Newton Iteration
        w = f(res[i-1][1], res[i-1][1])
        j = 0
        while abs(w0 - w) > eps:
            w0 = w
            w = f(w, res[i-1][1])
            j += 1
            if j >= iterations:
                print(f'Max iterations exceeded ({iterations})')
                return
        res.append((i*h, w))
    return res

In [9]:
# Actual Result
res = np.array([(0.1*i, math.exp(1-9*0.1*i)) for i in range(N+1)])
print(f'Actual Result:\n{res}')
print()

# Backward Euler
f = lambda x, x0: x - (1.9*x - x0)/1.9
w0 = math.e
backward_euler_res = np.array(backward_euler(w0, f))
print(f'Backward Euler Method:\n{backward_euler_res}')
print()
b_e_error = np.array([abs(res[i] - backward_euler_res[i]) for i in range(N+1)])
print(f'Backward Euler Error:\n{b_e_error}\nSum={sum(b_e_error)[1]}')
print()

# Euler
f = lambda t, w: -9*w
euler_res = np.array(euler(f, 0, 1, N=10, alpha=math.e))
print(f'Euler Method:\n{euler_res}')
print()
e_error = np.array([abs(res[i] - euler_res[i]) for i in range(N+1)])
print(f'Euler Error:\n{e_error}\nSum={sum(e_error)[1]}')
print()



Actual Result:
[[0.00000000e+00 2.71828183e+00]
 [1.00000000e-01 1.10517092e+00]
 [2.00000000e-01 4.49328964e-01]
 [3.00000000e-01 1.82683524e-01]
 [4.00000000e-01 7.42735782e-02]
 [5.00000000e-01 3.01973834e-02]
 [6.00000000e-01 1.22773399e-02]
 [7.00000000e-01 4.99159391e-03]
 [8.00000000e-01 2.02943064e-03]
 [9.00000000e-01 8.25104923e-04]
 [1.00000000e+00 3.35462628e-04]]

Backward Euler Method:
[[0.         2.71828183]
 [0.1        1.43067465]
 [0.2        0.75298666]
 [0.3        0.39630877]
 [0.4        0.20858356]
 [0.5        0.10978082]
 [0.6        0.05777938]
 [0.7        0.0304102 ]
 [0.8        0.01600537]
 [0.9        0.00842388]
 [1.         0.00443362]]

Backward Euler Error:
[[0.         0.        ]
 [0.         0.32550373]
 [0.         0.30365769]
 [0.         0.21362524]
 [0.         0.13430998]
 [0.         0.07958344]
 [0.         0.04550204]
 [0.         0.02541861]
 [0.         0.01397594]
 [0.         0.00759877]
 [0.         0.00409816]]
Sum=1.153273598353276


I am sorry for the unnecessary scientific notation, but it is still visible that, comparing the result of the Backward Euler Method and the Euler Method to the actual result, the Backward Euler Method is a lot closer to the actual result than the Euler Method. The Euler Method has larger absolute errors. Summing up all the errors from the data points, the Backward Euler Method has a total error of about 1.1533 whereas the Euler Method has a total error of about 4.2784, which is almost 3-4 times the error of the Backward Euler Method. A reason for this might be that this problem is laying outside of the region of stability of the Euler Method, which means that the absolute error of its results will keep increasing. Whereas the Backward Euler Method appears to be stable for this problem.